Adds NonBlocking Mode

madhurjain opened the pull request:
- Setting ReadTimeout to positive value sets non blocking mode
- ReadTimeout is specified in milliseconds irrespective of platform
- Caps ReadTimeout on Linux / POSIX systems
- Updated README

tarm commented on serial_posix.go:
```diff
@@ -77,7 +77,19 @@ func openPort(name string, baud int) (rwc io.ReadWriteCloser, err error) {
+	st.c_cc[C.VTIME] = C.cc_t(readTimeout)
+	
+	st.c_oflag &= ^C.tcflag_t(C.OPOST)
+	st.c_oflag &= ^C.tcflag_t(C.OPOST)
```
What are these two lines?  They look the same as line 79.  Why are they needed?

tarm commented on README.md:
````diff
@@ -58,6 +58,15 @@ func main() {
+	
+```go	
+	c := &serial.Config{Name: "COM45", Baud: 115200, NonBlockingRead: true, ReadTimeout: 1000}
+	// On Windows, ReadTimeout is in milliseconds (1000 = 1sec)
````
ReadTimeout should be a time.Duration and the os specific files should translate to their own units.

tarm commented on serial.go:
```diff
@@ -84,12 +86,13 @@ type Config struct {
 	// CRLFTranslate bool
-	// TimeoutStuff int
+	NonBlockingRead bool
+	ReadTimeout     uint32
```
This should be a time.Duration instead of a uint32 (see my comment in the README.md)

tarm commented on serial.go:
```diff
@@ -84,12 +86,13 @@ type Config struct {
 
 	// CRLFTranslate bool
-	// TimeoutStuff int
+	NonBlockingRead bool
```
Do we need this boolean?  Can we just special case the value of ReadTimeout = 0 and make that blocking?

tarm commented on serial_linux.go:
```diff
@@ -62,10 +62,15 @@ func openPort(name string, baud int) (rwc io.ReadWriteCloser, err error) {
 		Iflag:  syscall.IGNPAR,
 		Cflag:  syscall.CS8 | syscall.CREAD | syscall.CLOCAL | rate,
-		Cc:     [32]uint8{syscall.VMIN: 1},
+		Cc:     [32]uint8{syscall.VMIN: minBytesToRead, syscall.VTIME: uint8(readTimeout)},
```
Converting readTimeout to a uint8 will silently wrap values larger than 25.5 seconds.  Can you test for that case instead and return an error if the requested timeout is too long?

madhurjain replied:
Instead of returning an error, I thought it'd be better to cap the timeout to min / max

tarm replied:
OK, that sounds fine.

tarm commented:
I have a bunch of suggestions, but overall this seems like a reasonable approach.

What testing have you done?

madhurjain commented:
Thanks for your constructive comments
I have removed the `NonBlockingRead` boolean instead using positive value of `ReadTimeout` to set the port in non-blocking mode
Instead of using `time.Duration` type which internally stores value as int64 and does not have a Milliseconds function which would mean some conversions and complexity, I have tried to keep it simple by using uint32
I have also changed the `ReadTimeout` value to be in milliseconds irrespective of platform

I have tested it
- on Windows 7 (x32) Go 1.4 with Virtual Serial Port and a USB Modem
- on Ubuntu 14.04 (x64) Go 1.4 with `socat` and USB Modem
- on Mac OSX 10.10 Go 1.4.1 with `socat`

tarm commented on serial.go:
```diff
@@ -84,12 +85,12 @@ type Config struct {
 
 	// CRLFTranslate bool
-	// TimeoutStuff int
+	ReadTimeout uint32
```
I really think this should be a time.Duration.  time.Duration is the standard for intervals in Go and it has well understood units (nanoseconds) and conversion to milliseconds or 100 milliseconds is one easy operation.  For example, if this were to stay uint32 milliseconds, I think it should get an explicit "_ms" suffix and a comment saying the units.  time.Duration gets rid of that need.

Also please move this higher up in the struct and add a line of documentation about which kind of timeout it is.  My understanding of your intention is that this is an overall timeout NOT an interval timeout.  A comment would be very helpful to a user of the library.

tarm commented on serial_linux.go:
```diff
@@ -62,10 +62,26 @@ func openPort(name string, baud int) (rwc io.ReadWriteCloser, err error) {
 	fd := f.Fd()
+	// set blocking / non-blocking read
+	var minBytesToRead uint8 = 1
+	if readTimeout > 0 {
```
Can this logic that is duplicated twice be moved into a common shared function like

```
func posixTimeoutValues(readTimeout time.Duration) (vmin uint8, vtime uint8)
```

tarm commented on serial_linux.go:
```diff
@@ -62,10 +62,26 @@ func openPort(name string, baud int) (rwc io.ReadWriteCloser, err error) {
+			readTimeout = 25500
+		}
+		// convert milliseconds to deciseconds as expected by VTIME
+		readTimeout = readTimeout / 100
```
I think the intention and reason for the clamping would be clearer to do the /100 first and then clamp the values to 1 and math.MaxUint8.

tarm commented on serial_windows.go:
```diff
@@ -178,12 +178,20 @@ func setCommState(h syscall.Handle, baud int) error {
-	timeouts.ReadTotalTimeoutConstant = MAXDWORD - 1
+	if readTimeout > 0 {
+		// non-blocking read
+		timeouts.ReadIntervalTimeout = 1000
```
This make the inter-character timeout 1s, right?  I thought the timeout was for the overall read, not for characters?  It seems like this should be 0 instead.

tarm commented:
This is looking pretty good (just a few more nits).  I appreciate the cross-platform testing!

madhurjain commented:
Please check now.

tarm commented on serial.go:
```diff
@@ -84,12 +89,36 @@ type Config struct {
+	if readTimeout > 0 {
+		// EOF on zero read
+		minBytesToRead = 0
+		timeoutMs := uint32(readTimeout.Nanoseconds() / 1e6)
```
Why convert to a uint32?  That could cause wrapping.  I would leave it as an int64 like .Nanoseconds returns.  The bounds checks you have will work just the same.

tarm replied:
It seems like this could be simpler with something like

```
var readTimeoutInDeci int64
if readTimeout > 0 {
    readTimeoutInDeci = readTimeout.Nanoseconds() / 1e6 / 100
    if readTimeoutInDeci < 1 {
        readTimeoutInDeci = 1
    }
    if readTimeoutInDeci > MAXUINT8 {
        readTimeoutInDeci = MAXUINT8
    }
}

vtime = uint8(readTimeoutInDeci)
```

tarm commented on serial_windows.go:
```diff
@@ -178,12 +179,22 @@ func setCommState(h syscall.Handle, baud int) error {
-	timeouts.ReadIntervalTimeout = MAXDWORD
-	timeouts.ReadTotalTimeoutMultiplier = MAXDWORD
-	timeouts.ReadTotalTimeoutConstant = MAXDWORD - 1
+	timeoutMs := uint32(readTimeout.Nanoseconds() / 1e6)
```
Same comment about uint32 conversion here.

It would be better to keep it as an int64, then do the conversion to milliseconds, then do range checks, then do the conversion to uint32.

tarm commented on serial_windows.go:
```diff
@@ -178,12 +179,22 @@ func setCommState(h syscall.Handle, baud int) error {
-	timeouts.ReadTotalTimeoutConstant = MAXDWORD - 1
+	timeoutMs := uint32(readTimeout.Nanoseconds() / 1e6)
+
+	if timeoutMs > 0 {
```
There is a slight inconsistency between your implementation on the windows and the posix side:

If you ask for a ReadTimeout of 1 nanosecond, the posix implemention will give you non-blocking behavior with a timeout of 100ms.  The windows implementation will give you blocking behavior (because the division happens before the > 0 check).   The posix implementation seems slightly better, but really the important thing is consistency.

I might do it like this:

```
if readtimeTimeout > 0 {
    // non-blocking read
    timeoutMs := readTimeout.Nanoseconds() / 1e6
    if timeoutMs < 1 {
        timeoutMs = 1
    }
    if timeoutMs > math.MaxUint32 {
          timeoutMs = math.MaxUint32
    }
    timeouts.ReadIntervalTimeout = 0
    timeouts.ReadTotalTimeoutMultiplier = 0
    timeouts.ReadTotalTimeoutConstant = uint32(timeoutMs)
} else {
    // blocking read 
    ...
}
```

madhurjain replied:
That's bad. Thanks for pointing that out.
